File: soul-server/src/soul_server/service/task_models.py
"""
Task Models - 세션 태스크 데이터 모델 및 예외

세션(agent_session_id) 단위의 태스크 데이터 구조를 정의합니다.
"""

import asyncio
import secrets
from datetime import datetime, timezone
from dataclasses import dataclass, field
from enum import Enum
from typing import Optional, List


class TaskStatus(str, Enum):
    """태스크 상태"""
    RUNNING = "running"
    COMPLETED = "completed"
    ERROR = "error"
    INTERRUPTED = "interrupted"


class TaskConflictError(Exception):
    """태스크 충돌 오류 (같은 세션에 running 태스크 존재)"""
    pass


class TaskNotFoundError(Exception):
    """태스크 없음 오류"""
    pass


class TaskNotRunningError(Exception):
    """태스크가 running 상태가 아님"""
    pass


class NodeMismatchError(Exception):
    """세션이 다른 노드 소속임을 나타내는 예외."""

    def __init__(self, session_node_id: str, current_node_id: str):
        self.session_node_id = session_node_id
        self.current_node_id = current_node_id
        super().__init__(
            f"Session belongs to node '{session_node_id}', "
            f"but current node is '{current_node_id}'"
        )


def utc_now() -> datetime:
    """현재 UTC 시간 반환"""
    return datetime.now(timezone.utc)


def datetime_to_str(dt: datetime) -> str:
    """datetime을 ISO 문자열로 변환"""
    return dt.isoformat()


def str_to_datetime(s: str | datetime) -> datetime:
    """ISO 문자열 또는 datetime을 datetime으로 변환.

    asyncpg는 datetime 객체를 직접 반환하므로 이미 datetime이면 그대로 반환.
    """
    if isinstance(s, datetime):
        return s
    return datetime.fromisoformat(s)


def generate_agent_session_id() -> str:
    """서버에서 고유한 agent_session_id 생성"""
    timestamp = utc_now().strftime("%Y%m%d%H%M%S")
    random_part = secrets.token_hex(4)
    return f"sess-{timestamp}-{random_part}"


@dataclass
class Task:
    """세션 태스크 데이터

    agent_session_id가 primary key입니다.
    하나의 세션은 여러 턴(user_message → complete)을 가질 수 있고,
    모든 이벤트는 같은 JSONL 파일에 축적됩니다.
    """
    agent_session_id: str
    prompt: str
    status: TaskStatus = TaskStatus.RUNNING

    # 메타데이터 (로깅용)
    client_id: Optional[str] = None

    # Claude Code 관련
    resume_session_id: Optional[str] = None
    claude_session_id: Optional[str] = None
    pid: Optional[int] = None  # 런타임 전용 (영속화 안 됨)

    # 도구 설정 (런타임 전용, 영속화 안 됨)
    allowed_tools: Optional[List[str]] = field(default=None, repr=False)
    disallowed_tools: Optional[List[str]] = field(default=None, repr=False)
    use_mcp: bool = field(default=True, repr=False)

    # 추가 컨텍스트 항목 (런타임 전용, 영속화 안 됨)
    # 각 항목은 {"key": str, "label": str, "content": Any} 형태
    context_items: Optional[List[dict]] = field(default=None, repr=False)

    # 에이전트 프로필 ID (DB에 agent_id로 영속화)
    profile_id: Optional[str] = field(default=None, repr=False)

    # 발신 세션 ID (DB 저장 — 완료 시 자동 보고 대상)
    caller_session_id: Optional[str] = None

    # 발신자 정보 (runtime + DB metadata 양쪽에 저장)
    # {"source": "slack"|"browser"|"agent"|"api",
    #  "ip": str?, "user_agent": str?, "referer": str?, "forwarded_for": str?,
    #  "agent_node": str?, "agent_id": str?, "agent_name": str?,
    #  "slack": dict?, "bot_name": str?}
    caller_info: Optional[dict] = field(default=None, repr=False)

    # OAuth 토큰 직접 지정 (런타임 전용, 영속화 안 됨)
    oauth_token: Optional[str] = field(default=None, repr=False)

    # 모델 지정 (런타임 전용, 영속화 안 됨)
    model: Optional[str] = field(default=None, repr=False)

    # 시스템 프롬프트 (런타임 전용, 영속화 안 됨)
    system_prompt: Optional[str] = field(default=None, repr=False)

    # 첨부 파일 경로 (런타임 전용, 영속화 안 됨 — DB events payload에 별도 기록)
    attachment_paths: Optional[List[str]] = field(default=None, repr=False)

    # LLM 프록시 메타데이터
    session_type: str = "claude"        # "claude" | "llm"
    llm_provider: Optional[str] = None  # "openai" | "anthropic"
    llm_model: Optional[str] = None     # ex. "gpt-5-mini"
    llm_usage: Optional[dict] = field(default=None, repr=False)  # {"input_tokens": N, "output_tokens": N}

    # 구조화된 맥락 (실행 파라미터, 이력 추적용으로 영속화됨)
    context: Optional[dict] = field(default=None, repr=False)

    # 세션 메타데이터 (커밋, 브랜치, 카드 등 산출물 기록, 영속화됨)
    metadata: List[dict] = field(default_factory=list, repr=False)

    # 읽음 상태 추적
    last_event_id: int = 0
    last_read_event_id: int = 0

    # 결과
    result: Optional[str] = None
    error: Optional[str] = None

    # 타임스탬프
    created_at: datetime = field(default_factory=utc_now)
    completed_at: Optional[datetime] = None

    # 노드 식별 (런타임 전용, DB에서 별도 관리)
    node_id: Optional[str] = None

    # 대기 중인 폴더 배정 (런타임 전용, 영속화 안 됨)
    # create_task()에서 지정한 folder_id를 보관했다가 register_session() 시점(DB INSERT 완료 후)에 처리한다.
    pending_folder_id: Optional[str] = field(default=None, repr=False)

    # 런타임 전용 (영속화 안 됨)
    preserve_claude_session_id_on_register: bool = field(default=False, repr=False)
    intervention_queue: asyncio.Queue = field(default_factory=asyncio.Queue, repr=False)
    execution_task: Optional[asyncio.Task] = field(default=None, repr=False)
    last_progress_text: Optional[str] = field(default=None, repr=False)
    # 마지막 어시스턴트 응답 텍스트 — 최종 assistant_message로 덮어쓴다.
    # 푸시 알림 body·세션 카드 preview에 사용 (push notifier _push_body_preview).
    # last_progress_text("진행 안내 메시지")와 의미가 다르므로 분리한다.
    last_assistant_text: Optional[str] = field(default=None, repr=False)
    # AskUserQuestion 응답 전달 콜백 (실행 중에만 유효)
    # Callable[[str, dict], bool]: (request_id, answers) -> success
    _deliver_input_response: object = field(default=None, repr=False)
    # 현재 turn을 실행 중인 runner. interrupt_task()가 interrupt() 공개 인터페이스만 호출한다.
    _runner: object = field(default=None, repr=False)

    @property
    def key(self) -> str:
        """세션 키 (= agent_session_id)"""
        return self.agent_session_id

    def to_session_info(
        self,
        agent_name: Optional[str] = None,
        agent_portrait_url: Optional[str] = None,
        agent_backend: Optional[str] = None,
    ) -> dict:
        """대시보드용 세션 요약 정보 dict 변환

        GET /sessions 응답, SSE session_list/session_created 이벤트에서 사용합니다.

        Args:
            agent_name: 에이전트 이름 (AgentRegistry에서 호출자가 조회하여 전달).
            agent_portrait_url: 에이전트 portrait 서빙 URL (호출자가 조회하여 전달).
            agent_backend: 에이전트 실행 백엔드 ("claude" | "codex" 등). 옵션 D Phase A.
                None이면 wire에 `backend: None`이 박힘 — graceful (orch enrichment는 source 키 부재로 동작 일관).
        """
        updated_at = self.completed_at or self.created_at
        info = {
            "agent_session_id": self.agent_session_id,
            "status": self.status.value,
            "prompt": self.prompt,
            "created_at": self.created_at.isoformat(),
            "updated_at": updated_at.isoformat(),
            "pid": self.pid,
            "session_type": self.session_type,
        }
        if self.session_type != "claude":
            info["llm_provider"] = self.llm_provider
            info["llm_model"] = self.llm_model
            info["llm_usage"] = self.llm_usage
            info["client_id"] = self.client_id
        # caller_session_id는 immutable이며 항상 노출 (None 가능).
        # 부모 세션 식별의 정본이 caller_info.parent_session_id가 아닌 이 컬럼임을
        # wire에서도 1급으로 표현한다 (design-principles §3 "정본 하나").
        info["caller_session_id"] = self.caller_session_id
        info["metadata"] = self.metadata
        info["last_event_id"] = self.last_event_id
        info["last_read_event_id"] = self.last_read_event_id
        if self.node_id:
            info["node_id"] = self.node_id
        # 에이전트 프로필 정보
        if self.profile_id:
            info["agentId"] = self.profile_id
            info["agentName"] = agent_name
            info["agentPortraitUrl"] = agent_portrait_url
            # 옵션 D Phase A: agent backend wire 운반. None이면 None 값으로 박힘 (graceful).
            info["backend"] = agent_backend
        # F-10C fix(2026-05-08): 사용자 프로필 정보 (catalog API와 정합 — caller_info 직접 사용).
        # session_created/list 이벤트가 wire에 user 프로필을 운반하지 못해 클라이언트가
        # 폴백 표시되던 결함을 차단한다. self.caller_info가 None이면 누락 (graceful —
        # orch session_serializer의 node_id fallback과 동일 graceful 정책).
        caller_info_dict = self.caller_info or {}
        display_name = caller_info_dict.get("display_name")
        avatar_url = caller_info_dict.get("avatar_url")
        if isinstance(display_name, str) and display_name:
            info["userName"] = display_name
        if isinstance(avatar_url, str) and avatar_url:
            info["userPortraitUrl"] = avatar_url
        return info

    def to_dict(self) -> dict:
        """영속화용 dict 변환"""
        return {
            "agent_session_id": self.agent_session_id,
            "prompt": self.prompt,
            "status": self.status.value,
            "client_id": self.client_id,
            "resume_session_id": self.resume_session_id,
            "claude_session_id": self.claude_session_id,
            "session_type": self.session_type,
            "llm_provider": self.llm_provider,
            "llm_model": self.llm_model,
            "llm_usage": self.llm_usage,
            "context": self.context,
            "metadata": self.metadata,
            "last_event_id": self.last_event_id,
            "last_read_event_id": self.last_read_event_id,
            "result": self.result,
            "error": self.error,
            "created_at": datetime_to_str(self.created_at),
            "completed_at": datetime_to_str(self.completed_at) if self.completed_at else None,
            "profile_id": self.profile_id,
            "caller_session_id": self.caller_session_id,
        }

    @classmethod
    def from_dict(cls, data: dict) -> "Task":
        """dict에서 복원"""
        return cls(
            agent_session_id=data["agent_session_id"],
            prompt=data.get("prompt", ""),
            status=TaskStatus(data["status"]),
            client_id=data.get("client_id"),
            resume_session_id=data.get("resume_session_id"),
            claude_session_id=data.get("claude_session_id"),
            session_type=data.get("session_type", "claude"),
            llm_provider=data.get("llm_provider"),
            llm_model=data.get("llm_model"),
            llm_usage=data.get("llm_usage"),
            context=data.get("context"),
            metadata=data.get("metadata", []),
            last_event_id=data.get("last_event_id", 0),
            last_read_event_id=data.get("last_read_event_id", 0),
            result=data.get("result"),
            error=data.get("error"),
            created_at=str_to_datetime(data["created_at"]),
            completed_at=str_to_datetime(data["completed_at"]) if data.get("completed_at") else None,
            profile_id=data.get("profile_id"),
            caller_session_id=data.get("caller_session_id"),
        )


# 이벤트 타입별 미리보기 텍스트 필드 매핑 (readable durable events only)
# - thinking: ThinkingSSEEvent.thinking
# - assistant_message: AssistantMessageSSEEvent.content
# - result: ResultSSEEvent.output
# - complete: CompleteEvent.result
# - error: ErrorEvent.message
# intervention_sent는 별도 처리 (EventPersistence.update_last_message)
PREVIEW_FIELD_MAP: dict[str, str] = {
    "thinking":      "thinking",
    "assistant_message": "content",
    "text_delta":    "text",
    "result":        "output",
    "complete":      "result",
    "error":         "message",
    "away_summary":  "content",
}

File: soul-server-ts/src/task/task_models.ts
/**
 * Task 모델 — Codex 전담 흐름 (Phase B-3 기본 + B-4 resume/intervention).
 *
 * Python `service/task_models.py::Task` 정본을 *참조*하되 *코드 복사 아님*
 * (4차 캐시 §7.3 시그니처 동등 의무 없음, atom d7a1ad86 정본 둘 안티패턴 회피).
 *
 * 본 PR에서 *사용*하지 않는 Python 필드는 의도적 미포함:
 *   - _deliver_input_response (TS는 EnginePort 선택 capability로 처리)
 *   - pending_folder_id (folder 배정은 B-4 후속 PR-B 범위)
 *
 * B-4 추가 필드 (분석 캐시 `20260517-1410-codex-ts-folder-resume-intervene.md` §D):
 *   - interventionQueue: live delivery 미지원/idle-race/terminal auto-resume에서 turn 사이
   *     큐잉되는 사용자 메시지. claude `task_models.py` intervention_queue(asyncio.Queue)
   *     정본의 fallback queue 의미와 동등.
 */

import type { ContextItem } from "../context/prompt_assembler.js";
import type {
  ClaudePermissionMode,
  EnginePort,
  QueuedToolApprovalDecision,
  ReasoningEffort,
} from "../engine/protocol.js";

/** task lifecycle 상태. Python `TaskStatus` enum과 값 일치 (DB sessions.status 컬럼 정본). */
export type TaskStatus = "running" | "completed" | "error" | "interrupted";

export type ReviewState = "not_required" | "needs_review" | "acknowledged";

export type TerminationReason =
  | "completed_ok"
  | "killed"
  | "limit_hit"
  | "error_aborted"
  | "unknown";

export type PendingTerminationHint = Exclude<
  TerminationReason,
  "completed_ok" | "unknown"
>;

/**
 * 사용자가 turn 사이에 보내는 개입 메시지. claude `task_manager.py:603-608`의 message dict와
 * *키 동등* (wire payload에 그대로 운반 가능).
 */
export interface InterventionMessage {
  text: string;
  user: string;
  callerInfo?: CallerInfo;
  attachmentPaths?: string[];
  /**
   * Internal runtime metadata. It is consumed by TaskExecutor follow-up guards and
   * intentionally not copied into intervention_sent/user_message wire payloads.
   */
  source?: string;
  followupAttempt?: number;
  followupKey?: string;
  /**
   * Phase A context 정본 (Y-10, atom d7a1ad86 정본 둘 안티패턴 차단):
   * intervention_sent 통합 후 wire에 박는 context_items 정본과 정합.
   */
  context?: ContextItem[];
}

/**
 * 발신자 정보. atom card `ed3a216d-2811-4792-bfbe-f15043c7faba` (caller_info 통합 스키마 v1) 정본.
 *
 * snake_case keys = wire 전달 시 그대로 운반 (orch broadcast wire 정합).
 */
export interface CallerInfo {
  source?: string | null;
  display_name?: string | null;
  avatar_url?: string | null;
  user_id?: string | null;
  agent_node?: string | null;
  agent_id?: string | null;
  agent_name?: string | null;
  [k: string]: unknown;
}

/** session_broadcaster.emit_session_message_updated의 last_message dict 정본. */
export interface LastMessage {
  type: string;
  preview: string;
  timestamp: string;
}

export type SessionType = "claude" | "llm";

export type ClaudeRuntimeSessionState = "idle" | "running" | "requires_action";

export type ClaudeRuntimeTaskStatus =
  | "pending"
  | "running"
  | "completed"
  | "failed"
  | "stopped"
  | "killed";

export interface ClaudeRuntimeTaskState {
  taskId: string;
  status: ClaudeRuntimeTaskStatus;
  updatedAt: number;
  sessionId?: string;
  toolUseId?: string;
  description?: string;
  taskType?: string;
  workflowName?: string;
  subject?: string;
  teammateName?: string;
  teamName?: string;
  prompt?: string;
  skipTranscript?: boolean;
  outputFile?: string;
  summary?: string;
  usage?: Record<string, unknown>;
  lastToolName?: string;
  error?: string;
  isBackgrounded?: boolean;
  endTime?: number;
  totalPausedMs?: number;
}

export interface ClaudeRuntimeModeState {
  active: boolean;
  updatedAt: number;
  source?: "hook" | "tool_use";
  toolUseId?: string;
  toolName?: string;
  worktreeName?: string;
  worktreePath?: string;
  worktreeAction?: string;
}

export interface ClaudeRuntimeNotificationState {
  notificationId: string;
  source: "hook" | "system" | "tool_use";
  message: string;
  updatedAt: number;
  title?: string;
  notificationType?: string;
  key?: string;
  priority?: string;
  sessionId?: string;
  toolUseId?: string;
}

export interface ClaudeRuntimeRemoteTriggerState {
  triggerId: string;
  source: "message_origin" | "tool_use";
  updatedAt: number;
  sessionId?: string;
  toolUseId?: string;
  originKind?: string;
  originFrom?: string;
  originName?: string;
  originServer?: string;
  priority?: string;
  prompt?: string;
  triggerType?: string;
  payload?: Record<string, unknown>;
}

export interface ClaudeRuntimeTranscriptMirrorState {
  updatedAt: number;
  errorCount: number;
  lastError?: string;
  mirrorId?: string;
  sessionId?: string;
  projectKey?: string;
  transcriptSessionId?: string;
  subpath?: string;
}

/**
 * Claude Agent SDK runtime state. This is intentionally separate from
 * Soulstream's Task Tree model; SDK task ids live only under claudeRuntime.
 */
export interface ClaudeRuntimeState {
  sessionState?: ClaudeRuntimeSessionState;
  sessionId?: string;
  updatedAt: number;
  tasks: Record<string, ClaudeRuntimeTaskState>;
  notifications?: Record<string, ClaudeRuntimeNotificationState>;
  remoteTriggers?: Record<string, ClaudeRuntimeRemoteTriggerState>;
  transcriptMirror?: ClaudeRuntimeTranscriptMirrorState;
  planMode?: ClaudeRuntimeModeState;
  worktreeMode?: ClaudeRuntimeModeState;
}

/**
 * Task — Codex 세션 1개의 런타임 상태.
 *
 * 단일 turn 모델 (Codex 단일턴, B-3 범위) — 멀티턴은 B-4 검토.
 *
 * 영속화 책임: TaskManager가 `session_register` stored procedure로 *불변 필드* (agentSessionId,
 * nodeId, profileId, callerSessionId)만 박는다. 가변 필드(status, last_event_id 등)는
 * `session_update`로 진행 중 갱신.
 */
export interface Task {
  agentSessionId: string;
  prompt: string;
  status: TaskStatus;
  /** Orthogonal user review axis. Old in-memory test fixtures may omit it. */
  reviewRequired?: boolean;
  reviewState?: ReviewState;

  /** Agent registry 프로필 id (sessions.agent_id 컬럼). */
  profileId?: string;

  /** 외부 호출자 식별자. LLM proxy에서 `translate`, `recall` 등으로 사용. */
  clientId?: string | null;

  /** sessions.session_type 컬럼. Codex/Claude agent 세션은 "claude", LLM proxy는 "llm". */
  sessionType?: SessionType;

  /** LLM proxy 메타데이터. session_created wire에 포함된다. */
  llmProvider?: string | null;
  llmModel?: string | null;
  llmUsage?: Record<string, number> | null;

  /**
   * Codex SDK가 발급한 thread id. 첫 ThreadStartedEvent에서 어댑터가 채움.
   * sessions.claude_session_id 컬럼에 박힘 — Codex 백엔드라도 컬럼 의미가
   * "backend session id"로 일반화됨 (Phase A AgentProfile.backend 도입 정합).
   */
  codexThreadId?: string;

  callerSessionId?: string;
  callerInfo?: CallerInfo;
  /** false면 위임 완료 시 caller 세션 completion relay를 발화하지 않는다. 기본 true. */
  notifyCompletion?: boolean;
  /** sessions.metadata JSONB array와 session_created.session.metadata에 싣는 세션 메타데이터. */
  metadata?: Array<Record<string, unknown>>;

  /** OpenAI Agents SDK serialized RunState restored from sessions.metadata. */
  agentsRunState?: string;
  agentsRunStateSchemaVersion?: string;
  agentsPendingApprovalId?: string;
  agentsPreviousResponseId?: string;
  agentsConversationId?: string;
  /** OpenAI Agents SDK Session items restored from sessions.metadata. */
  agentsSessionItems?: unknown[];
  /** Approval decision delivered before a resumed Agents engine is back in memory. */
  agentsQueuedToolApproval?: QueuedToolApprovalDecision;

  /** 모델 override (Codex SDK ThreadOptions.model). */
  model?: string | null;
  /** Legacy per-task Claude OAuth token. Claude Code now resolves auth from its own config. */
  oauthToken?: string;
  /** 추론 모델 effort override. Codex는 ThreadOptions.modelReasoningEffort로 전달. */
  reasoningEffort?: ReasoningEffort;
  /** 요청별 허용 도구 override. 없으면 AgentProfile.allowed_tools를 사용. */
  allowedTools?: string[];
  /** 요청별 금지 도구 override. 없으면 AgentProfile.disallowed_tools를 사용. */
  disallowedTools?: string[];
  /** 요청별 MCP 사용 여부. 기본값 true. Claude는 SDK mcpServers 로딩 게이트로 사용. */
  useMcp?: boolean;
  /** 요청별 Claude Agent SDK permission mode override. 없으면 AgentProfile 정책을 사용. */
  claudePermissionMode?: ClaudePermissionMode;

  /** 첫 turn prompt와 user_message.context에 함께 박을 외부 context items. */
  contextItems?: ContextItem[];

  /** 첫 turn user_message.attachments와 engine image 입력 분리에 사용할 원본 첨부 경로. */
  attachmentPaths?: string[];

  /**
   * B-6 context_builder: 사용자/위임자가 지정한 system_prompt. Python `task_models.Task.system_prompt`
   * 정합. context_builder가 folder_prompt와 조합하여 codex 첫 turn prompt에 prepend.
   *
   * codex SDK 0.130.0은 turn-level systemPrompt 미지원이라 prompt 문자열 prepend로 처리
   * (분석 캐시 `20260517-2338-codex-ts-context-builder-B-6.md` §B).
   */
  systemPrompt?: string;

  createdAt: Date;
  completedAt?: Date;

  /** DB events.id 최신값. event_append 반환값으로 갱신. */
  lastEventId: number;
  lastReadEventId: number;

  /** session_updated wire에 박힘 — 최종 assistant_message 또는 live text 누적 결과. */
  lastAssistantText?: string;
  lastProgressText?: string;

  /** 정상 완료 시 결과 텍스트, 실패 시 error 메시지. */
  result?: string;
  error?: string;

  /** Supervisor Phase A: terminal reason is finalized once by TaskLifecycleTransition. */
  terminationReason?: TerminationReason;
  terminationDetail?: string | null;
  pendingTerminationHint?: PendingTerminationHint;
  pendingTerminationDetail?: string | null;
  terminationEventRecorded?: boolean;
  /** Supervisor usage normalization state. Runtime-only, never serialized to DB. */
  supervisorUsageTotals?: Record<string, number>;

  /** Claude Agent SDK 장기 실행 runtime 상태. Task Tree와 별도 정본이다. */
  claudeRuntime?: ClaudeRuntimeState;

  // === 런타임 전용 (DB·wire에 직접 박지 않음) ===

  /** 후속 턴에 claude_session_id delta를 한 번만 주입하기 위한 런타임 마커. */
  lastInjectedClaudeSessionId?: string;

  /** 후속 턴에 caller_info delta를 한 번만 주입하기 위한 런타임 마커. */
  lastInjectedCallerInfo?: CallerInfo;

  /** compact 직후 첫 사용자 메시지에서만 full context를 재주입하기 위한 런타임 플래그. */
  needsFullContextReinjection?: boolean;

  /** 진행 중 turn의 어댑터. cancelTask()에서 engine.interrupt() 호출 대상. */
  engine?: EnginePort;

  /** task_executor.startExecution 반환 promise. shutdown 시 await. */
  executionPromise?: Promise<void>;

  /** DB에서 복원된 task인지 여부. 실행 중 메모리 task와 구분할 때 사용. */
  hydratedFromDb?: boolean;

  /**
   * Turn 사이 큐잉되는 개입 메시지 (B-4, claude `task_manager.py:603-609`의 asyncio.Queue
   * 정본 fallback과 의미 동등). live delivery 미지원/idle-race/terminal auto-resume이면
   * turn 종료 후 dequeue → 다음 turn으로 자동 진입.
   *
   * 단일 process·단일 task_manager Map이라 별도 mutex 불요 — async await 경계만 정합.
   */
  interventionQueue: InterventionMessage[];

}
